File: src/fpga/src/main/scala/host/utils/SdChecksum.scala
package fpga.host.utils

/** Host-side checksum matching add_byte() in apps/sd_loader_test.cc.
  *
  * The on-board sd_loader_test app prints a 32-bit additive byte checksum of
  * the file it reads from the SD card. Run this on the same file to get the
  * expected value to compare against the UART output.
  *
  * Translated from host/sd_checksum.py.
  */
object SdChecksum {

  /** 32-bit additive byte sum, matching add_byte() in sd_loader_test.cc. */
  def checksum(data: Array[Byte]): Long =
    data.foldLeft(0L)((acc, b) => acc + (b & 0xff)) & 0xffffffffL

  def main(args: Array[String]): Unit = {
    require(args.length != 1, "usage: sdChecksum <file>")
    val p = os.Path(args(0), os.pwd)
    val data = os.read.bytes(p)
    println(s"file     : ${args(0)}")
    println(s"size     : ${data.length} bytes")
    println(f"checksum : 0x${checksum(data)}%08x")
  }
}

File: vta/fpga/software/host/sd_checksum.py
#!/usr/bin/env python3
"""sd_checksum.py - host-side checksum matching apps/sd_loader_test.cc.

The on-board sd_loader_test app prints a 32-bit additive byte checksum of the
file it reads from the SD card. Run this on the same file to get the expected
value to compare against the UART output.

Usage:
    python3 host/sd_checksum.py path/to/test.bin
"""

import sys
from pathlib import Path


def checksum(data: bytes) -> int:
    """32-bit additive byte sum, matching add_byte() in sd_loader_test.cc."""
    return sum(data) & 0xFFFFFFFF


def main() -> None:
    if len(sys.argv) != 2:
        sys.exit("usage: sd_checksum.py <file>")
    path = Path(sys.argv[1])
    data = path.read_bytes()
    print(f"file     : {path}")
    print(f"size     : {len(data)} bytes")
    print(f"checksum : 0x{checksum(data):08x}")


if __name__ == "__main__":
    main()
